File: runtimes/ollama/tools.go
// Allowlist composition — a port of runtimes/claude/tools.js — plus the gate
// and the registry over both tool sources.
//
// The manager sends HALF the allowlist (the wiring's toolsets) and the mode;
// the other half is the `tools:` frontmatter of .claude/agents/<agent>.md in
// the checkout, which only this process can read.
//
// Deliberately NOT a YAML parser. It reads one field of one shape and treats
// everything it does not understand as "declares nothing" — an unreadable role
// file must never stop an agent from answering.
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

const (
	ModeMerge     = "merge"
	ModeOverwrite = "overwrite"
)

// splitList turns a comma-separated allowlist into trimmed entries.
func splitList(s string) []string {
	var out []string
	for _, t := range strings.Split(s, ",") {
		if t = strings.TrimSpace(t); t != "" {
			out = append(out, t)
		}
	}
	return out
}

func unquote(s string) string {
	t := strings.TrimSpace(s)
	if len(t) >= 2 && ((t[0] == '"' && t[len(t)-1] == '"') || (t[0] == '\'' && t[len(t)-1] == '\'')) {
		return strings.TrimSpace(t[1 : len(t)-1])
	}
	return t
}

var (
	reToolsKey  = regexp.MustCompile(`^tools:(.*)$`)
	reBlockItem = regexp.MustCompile(`^\s+-\s*(.*)$`)
)

// parseFrontmatterTools extracts `tools:` from an agent definition's YAML
// frontmatter. An absent key or absent frontmatter yields an empty list with a
// nil error; a file that was there but could not be understood returns an
// error so it can be LOGGED rather than pass silently.
func parseFrontmatterTools(text string) ([]string, error) {
	text = strings.TrimPrefix(text, "\ufeff")
	lines := strings.Split(strings.ReplaceAll(text, "\r\n", "\n"), "\n")
	i := 0
	for i < len(lines) && strings.TrimSpace(lines[i]) == "" {
		i++
	}
	if i >= len(lines) || strings.TrimSpace(lines[i]) != "---" {
		return []string{}, nil
	}
	start, end := i+1, -1
	for j := start; j < len(lines); j++ {
		if t := strings.TrimSpace(lines[j]); t == "---" || t == "..." {
			end = j
			break
		}
	}
	if end < 0 {
		return nil, errors.New("frontmatter opened with --- but never closed")
	}
	for j := start; j < end; j++ {
		m := reToolsKey.FindStringSubmatch(lines[j])
		if m == nil {
			continue // top-level key only: an indented tools: belongs to something else
		}
		inline := strings.TrimSpace(m[1])
		if strings.HasPrefix(inline, "[") {
			if !strings.HasSuffix(inline, "]") {
				return nil, errors.New("tools: flow list is not closed on one line")
			}
			return unquoteAll(splitList(inline[1 : len(inline)-1])), nil
		}
		if inline != "" {
			return unquoteAll(splitList(inline)), nil
		}
		tools := []string{}
		for k := j + 1; k < end; k++ {
			if strings.TrimSpace(lines[k]) == "" {
				continue
			}
			item := reBlockItem.FindStringSubmatch(lines[k])
			if item == nil {
				break
			}
			if v := unquote(item[1]); v != "" {
				tools = append(tools, v)
			}
		}
		return tools, nil
	}
	return []string{}, nil
}

func unquoteAll(in []string) []string {
	out := []string{}
	for _, s := range in {
		if v := unquote(s); v != "" {
			out = append(out, v)
		}
	}
	return out
}

// agentDeclaredTools reads .claude/agents/<agent>.md under workspace. Every
// failure yields an empty list; the run continues either way. logf receives a
// one-line reason when something was there but could not be used.
func agentDeclaredTools(workspace, agent string, logf func(string, ...any)) []string {
	if workspace == "" || agent == "" {
		return nil
	}
	text, err := os.ReadFile(filepath.Join(workspace, ".claude", "agents", agent+".md"))
	if err != nil {
		return nil
	}
	tools, perr := parseFrontmatterTools(string(text))
	if perr != nil {
		logf("[runtime] agent definition %s.md: %v — treating it as declaring no tools", agent, perr)
		return nil
	}
	return tools
}

// composeAllowedTools joins the two halves per mode. Any mode not recognised —
// including an absent one — is merge, because reading it as overwrite would
// silently strip what the agent declared.
func composeAllowedTools(agentTools []string, wiring string, mode string) []string {
	w := splitList(wiring)
	if mode == ModeOverwrite {
		return dedup(w)
	}
	return dedup(append(append([]string{}, agentTools...), w...))
}

func dedup(list []string) []string {
	seen := map[string]bool{}
	out := []string{}
	for _, raw := range list {
		t := strings.TrimSpace(raw)
		if t == "" || seen[t] {
			continue
		}
		seen[t] = true
		out = append(out, t)
	}
	return out
}

// patternMatches decides whether an allowlist entry grants a tool name.
//
// Exact name, a trailing `*` prefix, and therefore `mcp__server__*`. A
// claude-style narrowing specifier such as `Bash(kubectl:*)` grants NOTHING:
// reading it as bare Bash would widen a grant the operator wrote to narrow it.
// The caller logs those, once, so a binding that means nothing here is visible.
func patternMatches(pattern, name string) bool {
	if strings.Contains(pattern, "(") {
		return false
	}
	if strings.HasSuffix(pattern, "*") {
		return strings.HasPrefix(name, strings.TrimSuffix(pattern, "*"))
	}
	return pattern == name
}

// isNarrowing reports a specifier this runtime cannot honour.
func isNarrowing(pattern string) bool { return strings.Contains(pattern, "(") }

// Tool is one executable tool, from either source.
type Tool struct {
	Name        string
	Description string
	Schema      json.RawMessage
	// Call executes it. A tool ERROR is returned as (text, true, nil) so the
	// model can read and recover from it; a nil error with isError=false is a
	// result. A non-nil error is the tool layer itself failing.
	Call func(ctx context.Context, args json.RawMessage) (string, bool, error)
}

// Registry holds every tool this pod can provide, keyed by name.
type Registry struct {
	tools map[string]Tool
	order []string
}

func NewRegistry() *Registry { return &Registry{tools: map[string]Tool{}} }

func (r *Registry) Add(t Tool) {
	if _, dup := r.tools[t.Name]; !dup {
		r.order = append(r.order, t.Name)
	}
	r.tools[t.Name] = t
}

func (r *Registry) Get(name string) (Tool, bool) { t, ok := r.tools[name]; return t, ok }

// Gate applies the allowlist ONCE, before the request: the returned tools are
// the only ones advertised. It also reports what the allowlist named that
// nothing here provides — per the catalog rule, a runtime provides what it can
// and REPORTS what it cannot — and every narrowing specifier it ignored.
func (r *Registry) Gate(allowed []string) (granted []Tool, unavailable []string, ignored []string) {
	seen := map[string]bool{}
	for _, p := range allowed {
		if isNarrowing(p) {
			ignored = append(ignored, p)
			continue
		}
		matched := false
		for _, name := range r.order {
			if patternMatches(p, name) {
				matched = true
				if !seen[name] {
					seen[name] = true
					granted = append(granted, r.tools[name])
				}
			}
		}
		if !matched {
			unavailable = append(unavailable, p)
		}
	}
	return granted, unavailable, ignored
}

func toolDefs(tools []Tool) []ToolDef {
	out := make([]ToolDef, 0, len(tools))
	for _, t := range tools {
		out = append(out, newToolDef(t.Name, t.Description, t.Schema))
	}
	return out
}

// toolErr formats a readable error for the model.
func toolErr(format string, a ...any) (string, bool, error) {
	return "error: " + fmt.Sprintf(format, a...), true, nil
}

File: runtimes/claude/tools.js
// Tool-allowlist composition for the claude runtime.
//
// The manager sends HALF the allowlist — the tools the conversation's wiring
// (its Pipeline's MCPToolsets) contributes — plus the mode saying how it
// composes with the other half: what the agent's own definition declares in
// the `tools:` frontmatter of .claude/agents/<agent>.md. Only the runtime holds
// the repository, so only the runtime can read that half and finish the job.
//
// Verified against the real CLI (claude 2.1.226): a definition's `tools:` does
// not widen or narrow the main session on its own — `--allowedTools` is the
// sole permission authority. Nothing composes these two unless we do it here.
//
// Deliberately NOT a YAML parser. It reads one field of one shape and treats
// everything it does not understand as "declares nothing" — an unreadable role
// file must never stop an agent from answering.

'use strict';

const fs = require('fs');
const path = require('path');

const MODE_MERGE = 'merge';
const MODE_OVERWRITE = 'overwrite';

// safeJoin resolves base/...segments and refuses a result that escapes base
// -- jssecurity:S2083's ask, since the caller below joins a directory this
// process controls with a name that arrives from data it does not (a CR's
// agent name); runtime.js's promptFile caller does the same for a work
// unit's prompt file. null on escape, so a caller can treat it exactly like
// a missing file rather than reading outside its intended tree.
function safeJoin(base, ...segments) {
  const root = path.resolve(base);
  const target = path.resolve(root, ...segments);
  if (target !== root && !target.startsWith(root + path.sep)) return null;
  return target;
}

// sanitizeLog strips CR/LF from a value before it reaches a log line --
// jssecurity:S5145's ask, since a crafted runId, agent name or thread id in a
// work unit could otherwise forge a second log line that reads as the
// runtime's own. /[\n\r]/g is the rule's own documented compliant pattern.
function sanitizeLog(v) {
  return String(v).replace(/[\n\r]/g, '_');
}

// resolveBin looks `name` up on PATH once and returns the first absolute
// hit, falling back to the bare name so a missing binary still fails with
// the ordinary ENOENT a reader expects. go:S4036's ask, one process over:
// say explicitly where a spawned binary comes from, rather than leaving
// PATH to answer it implicitly at the call site.
function resolveBin(name, pathEnv = process.env.PATH || '') {
  for (const dir of pathEnv.split(path.delimiter)) {
    if (!dir) continue;
    const candidate = path.join(dir, name);
    try {
      fs.accessSync(candidate, fs.constants.X_OK);
      return candidate;
    } catch {
      // not here — keep looking
    }
  }
  return name;
}

// splitList turns a comma/newline separated allowlist into trimmed entries.
function splitList(s) {
  if (!s) return [];
  return String(s).split(',').map((t) => t.trim()).filter(Boolean);
}

// unquote strips one layer of matching quotes from a scalar.
function unquote(s) {
  const t = s.trim();
  if (t.length >= 2 && ((t[0] === '"' && t.endsWith('"')) || (t[0] === "'" && t.endsWith("'")))) {
    return t.slice(1, -1).trim();
  }
  return t;
}

// parseFrontmatterTools extracts the `tools:` declaration from an agent
// definition's YAML frontmatter.
//
// Returns { tools: string[] } when it understood the file (an absent `tools:`
// key yields an empty list), or { error: <reason> } when it did not. Callers
// treat both as "contributes nothing"; the error only exists so it can be
// logged rather than passing silently.
function parseFrontmatterTools(text) {
  const lines = String(text).replace(/^﻿/, '').split(/\r?\n/);
  let i = 0;
  while (i < lines.length && lines[i].trim() === '') i++;
  if (i >= lines.length || lines[i].trim() !== '---') {
    return { tools: [] }; // no frontmatter at all — declares nothing, not an error
  }
  const start = i + 1;
  let end = -1;
  for (let j = start; j < lines.length; j++) {
    const t = lines[j].trim();
    if (t === '---' || t === '...') { end = j; break; }
  }
  if (end < 0) return { error: 'frontmatter opened with --- but never closed' };

  for (let j = start; j < end; j++) {
    const m = /^tools:(.*)$/.exec(lines[j]);
    if (!m) continue; // top-level key only: an indented `tools:` belongs to something else
    const inline = m[1].trim();

    if (inline.startsWith('[')) {
      if (!inline.endsWith(']')) return { error: 'tools: flow list is not closed on one line' };
      return { tools: splitList(inline.slice(1, -1)).map(unquote).filter(Boolean) };
    }
    if (inline !== '') {
      return { tools: splitList(inline).map(unquote).filter(Boolean) };
    }
    // block form: the following indented "- item" lines
    const tools = [];
    for (let k = j + 1; k < end; k++) {
      const raw = lines[k];
      if (raw.trim() === '') continue;
      // No adjacent quantifiers (javascript:S8786): the leading indent, the
      // dash and the item text are each matched by their own single-quantifier
      // step instead of one regex where \s* and .* could both claim the same
      // characters.
      const indent = /^\s+/.exec(raw);
      if (!indent || raw[indent[0].length] !== '-') break; // next key — the block ended
      const v = unquote(raw.slice(indent[0].length + 1).replace(/^\s+/, ''));
      if (v) tools.push(v);
    }
    return { tools };
  }
  return { tools: [] }; // frontmatter present, no tools: key — declares nothing
}

// agentDeclaredTools reads .claude/agents/<agent>.md under workspace and
// returns what it declares. Absent file, absent frontmatter, absent `tools:`,
// and unparseable frontmatter all yield [] — the run continues either way.
// log is called with a one-line reason whenever something was there but could
// not be used, so a typo in a role file is visible in the pod log.
function agentDeclaredTools(workspace, agent, log = () => {}) {
  if (!workspace || !agent) return [];
  const file = safeJoin(workspace, '.claude', 'agents', `${agent}.md`);
  if (!file) {
    log(`[runtime] agent definition ${agent}.md: path escapes the workspace — treating it as declaring no tools`);
    return [];
  }
  let text;
  try {
    text = fs.readFileSync(file, 'utf8');
  } catch {
    return []; // no definition — the wiring's tools stand alone
  }
  const parsed = parseFrontmatterTools(text);
  if (parsed.error) {
    log(`[runtime] agent definition ${agent}.md: ${parsed.error} — treating it as declaring no tools`);
    return [];
  }
  return parsed.tools;
}

// composeAllowedTools joins the agent's declared tools with the wiring's per
// mode: overwrite passes the wiring's alone, merge unions them with the
// agent's keeping their position. Any mode we do not recognise — including an
// absent one, which is what an object stored before the field existed sends —
// is merge, because reading it as overwrite would silently strip what the
// agent declared.
function composeAllowedTools(agentTools, wiringTools, mode) {
  const wiring = Array.isArray(wiringTools) ? wiringTools : splitList(wiringTools);
  if (mode === MODE_OVERWRITE) return dedup(wiring);
  const agent = Array.isArray(agentTools) ? agentTools : splitList(agentTools);
  return dedup(agent.concat(wiring));
}

function dedup(list) {
  const seen = new Set();
  const out = [];
  for (const raw of list) {
    const t = String(raw).trim();
    if (!t || seen.has(t)) continue;
    seen.add(t);
    out.push(t);
  }
  return out;
}

// buildClaudeArgs is the CLI invocation's argv, pulled out of runtime.js so
// its exact shape -- in particular the jssecurity:S6350 fix -- is asserted
// directly rather than only exercised end to end. `contextId` and `prompt`
// both arrive from a work unit and are values a naive `--flag value`
// pairing hands the CLI's parser openly: a prompt beginning with "-" is
// misread as an unrecognised OPTION ("error: unknown option ..."), verified
// against the real CLI, and `--resume` takes an OPTIONAL argument ([value]
// in --help) so the same misparse hits a dash-leading context id too, one
// path over -- the value reads as absent and the id is parsed as its own
// token. `--resume=<id>` (one `=`-joined token) and a `--` separator
// immediately before the trailing positional `prompt` are the fixes;
// `--append-system-prompt <value>` needs neither, since a REQUIRED-argument
// option already consumes the very next token unconditionally.
function buildClaudeArgs({ contextId, systemPrompt, allowed, maxTurns, mcpConfig, prompt }) {
  return [
    ...(contextId ? [`--resume=${contextId}`] : []),
    ...(systemPrompt ? ['--append-system-prompt', systemPrompt] : []),
    '-p',
    '--allowedTools', allowed.join(','),
    '--permission-mode', 'dontAsk',
    '--max-turns', String(maxTurns || 60),
    '--output-format', 'stream-json',
    '--verbose',
    '--strict-mcp-config',
    '--mcp-config', mcpConfig,
    '--', prompt,
  ];
}

module.exports = {
  MODE_MERGE, MODE_OVERWRITE,
  splitList, parseFrontmatterTools, agentDeclaredTools, composeAllowedTools,
  safeJoin, sanitizeLog, resolveBin, buildClaudeArgs,
};
